src/Numerics/SystemSolvers/batched_generalized_minimal_residual_solver.jl

```diff
@@ -263,517 +276,381 @@ function BatchedGeneralizedMinimalResidual(
-    gmres.cs[1 + 2 * (n - 1), I] = gmres.R[n, n, I]
-    gmres.cs[2 * n, I] = gmres.H[n + 1, I]
-    gmres.R[n, n, I] =
-        sqrt(gmres.cs[1 + 2 * (n - 1), I]^2 + gmres.cs[2 * n, I]^2)
-    gmres.cs[1 + 2 * (n - 1), I] /= gmres.R[n, n, I]
-    gmres.cs[2 * n, I] /= -gmres.R[n, n, I]
-    return nothing
-end
 
-"""
-    solve_optimization!(iteration, gmres, I)
-
-# Description
-Solves the optimization problem in GMRES
-# Arguments
-- `iteration`: (int) current iteration number
-- `gmres`: (struct) [OVERWRITTEN]
-- `I`: (int) thread index
-# Return
-nothing
-"""
-@inline function solve_optimization!(n, gmres, I)
-    # just need to update rhs from previous iteration
-    # apply latest givens rotation
-    tmp1 =
-        gmres.cs[1 + 2 * (n - 1), I] * gmres.rhs[n, I] -
-        gmres.cs[2 * n, I] * gmres.rhs[n + 1, I]
-    gmres.rhs[n + 1, I] =
-        gmres.cs[2 * n, I] * gmres.rhs[n, I] +
-        gmres.cs[1 + 2 * (n - 1), I] * gmres.rhs[n + 1, I]
-    gmres.rhs[n, I] = tmp1
-    # gmres.rhs[iteration+1] is the residual. Technically convergence should be checked here.
-    gmres.residual[n, I] = abs.(gmres.rhs[n + 1, I])
-    # copy for performing the backsolve and saving gmres.rhs
-    @inbounds for i in 1:n
-        gmres.sol[i, I] = gmres.rhs[i, I]
-    end
-    # do the backsolve
-    @inbounds for i in n:-1:1
-        gmres.sol[i, I] /= gmres.R[i, i, I]
-        @inbounds for j in 1:(i - 1)
-            gmres.sol[j, I] -= gmres.R[j, i, I] * gmres.sol[i, I]
+    # Having determined yᵢ, we now construct the GMRES solution
+    # in each column: xⱼ = ∑ᵢ yᵢ Ψᵢ
+    @inbounds for i in 1:j
+        @inbounds for k in 1:Ndof
+            sols[k, cidx] += g0s[cidx, i] * batched_krylov_basis[i, k, cidx]
         end
     end
-    return nothing
+
+    nothing
 end
 
 """
-    compute_residuals(gmres, i)
+    convert_structure!(
+        x,
+        y,
+        reshape_tuple,
+        permute_tuple,
+    )
 
-# Description
-Compute atol and rtol of current iteration
+Computes a tensor transpose and stores result in x
 
 # Arguments
-- `gmres`: (struct)
-- `i`: (current iteration)
-
-# Return
-- `atol`: (float) absolute tolerance
-- `rtol`: (float) relative tolerance
-
-# Comment
-sometimes gmres.R[1, 1,:] term has some very large components which makes rtol quite small
+- `x`: (array) [OVERWRITTEN]. target destination for storing the y data
+- `y`: (array). data that we want to copy
+- `reshape_tuple`: (tuple) reshapes y to be like that of x, up to a permutation
+- `permute_tuple`: (tuple) permutes the reshaped array into the correct structure
 """
-function compute_residuals(gmres, i)
-    atol = maximum(gmres.residual[i, :])
-    rtol = maximum(gmres.residual[i, :] ./ norm(gmres.R[1, 1, :]))
-    return atol, rtol
+@inline function convert_structure!(x, y, reshape_tuple, permute_tuple)
+    alias_y = reshape(y, reshape_tuple)
+    permute_y = permutedims(alias_y, permute_tuple)
+    copyto!(x, reshape(permute_y, size(x)))
+    nothing
+end
```

Comment:
I still don't understand why we need this copying of the data. It seems that we should be able to have the kernels be aware of the underlying data layout we are using.

Reply:
I completely agree. That will be addressed separately!